File: apps/ios/Voiid/Voiid/Games/Ludo/LudoPlayerPod.swift
//
//  LudoPlayerPod.swift
//  Voiid
//
//  A player pod (§11.2): EXACTLY two visible elements — the timer-ring/color-chip assembly at
//  its outer edge and one username line. The ring SURROUNDS the chip rather than adding a third
//  ornament. No avatar, badge, level, rank, gift, coin, completion dots, scores or status icons;
//  finished pawns stay visible in the center triangle instead.
//
//  USERNAME RULES: one line, tail-truncated at 18 clusters, never below 12pt, never marquee.
//  The viewer's own pod shows their username, not "You". Waiting seat: outlined chip +
//  "Waiting…". Dropped seat: keeps position, desaturated chip, name at 55%, no timer.
//

import SwiftUI

struct LudoPlayerPod: View {
    let seatView: LudoSeatViewV2?
    let active: Bool
    /// 0…1 remaining while THIS seat's decision window is open; nil otherwise.
    let ringFraction: Double?
    /// Warning/critical override for the arc; nil → the player hue.
    let ringColorOverride: Color?
    var compact = false
    var accessibilityLabel: String = ""

    @Environment(\.colorScheme) private var scheme

    var body: some View {
        let colors = LudoColors.resolve(scheme)
        let size = compact ? LudoDimens.podSizeCompact : LudoDimens.podSizeStandard
        let ring = compact ? LudoDimens.ringCompact : LudoDimens.ringStandard
        let chip = compact ? LudoDimens.chipCompact : LudoDimens.chipStandard

        HStack(spacing: 6) {
            ZStack {
                // Ring surrounds the color chip — never a third ornament (§11.2).
                LudoTimerRing.drawView(
                    diameter: ring,
                    stroke: LudoDimens.ringStroke,
                    track: colors.timerTrack,
                    arc: arcColor(colors),
                    fraction: active ? (ringFraction ?? 1) : nil)
                    .frame(width: ring, height: ring)
                Circle()
                    .strokeBorder(outlined ? colors.yardPocketBorder : .clear, lineWidth: 1)
                    .background(Circle().fill(chipColor(colors)))
                    .frame(width: chip, height: chip)
            }
            .padding(.leading, (size.height - ring) / 2)

            Text(usernameLine)
                .font(.system(size: 12, weight: .semibold, design: .rounded))
                .lineLimit(1)
                .truncationMode(.tail)
                .foregroundStyle(colors.textPrimary)
                .frame(maxWidth: .infinity, alignment: .leading)
                .padding(.trailing, 6)
        }
        .frame(width: size.width, height: size.height)
        .background(
            RoundedRectangle(cornerRadius: LudoDimens.podCornerRadius)
                .fill(colors.podSurface))
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(accessibilityLabel)
    }

    private var usernameLine: String {
        guard let sv = seatView else { return "" }
        if sv.participation == "waiting" { return "Waiting…" }
        let name = String(sv.displayName.prefix(18))
        return sv.isBot ? "\(name) BOT" : name
    }

    private var outlined: Bool {
        seatView == nil || seatView?.participation == "waiting"
    }

    private func chipColor(_ c: LudoColors) -> Color {
        guard let sv = seatView else { return .clear }
        return outlined ? .clear : c.hue(sv.seat)
    }

    private func arcColor(_ c: LudoColors) -> Color {
        guard active else { return c.timerTrack }          // inactive pods draw only the track
        if let o = ringColorOverride { return o }
        return c.timerActive
    }
}

private extension LudoTimerRing {
    /// View wrapper so pods can drop the ring straight into a frame. Inactive pods draw ONLY
    /// the neutral track circle; an active pod replaces that track progressively with the
    /// depleting arc (§11.2).
    static func drawView(diameter: CGFloat, stroke: CGFloat,
                         track: Color, arc: Color, fraction: Double?) -> some View {
        Canvas { ctx, size in
            var ctx = ctx
            draw(&ctx, diameter: min(size.width, size.height), stroke: stroke,
                 track: track,
                 arc: fraction != nil ? arc : .clear,
                 fraction: fraction ?? 0)
        }
        .frame(width: diameter, height: diameter)
    }
}

File: apps/android/app/src/main/java/com/voiid/app/main/games/ludo/LudoPlayerPod.kt
package com.voiid.app.main.games.ludo

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.graphics.Color
import com.voiid.app.ui.theme.LudoDimens
import com.voiid.app.ui.theme.LudoPalette
import com.voiid.app.ui.theme.ludoPaletteFor

/**
 * A player pod (§11.2): EXACTLY two visible elements — the timer-ring/color-chip assembly at
 * its outer edge and one username line. No avatar, badge, level, rank, gift, coin, completion
 * dots, scores or status icons; finished pawns stay visible in the center triangle instead.
 *
 * USERNAME RULES: one line, tail-truncated at 18 clusters, never below 12sp, never marquee.
 * The viewer's own pod shows their username, not "You". A waiting seat shows an outlined chip
 * and "Waiting…". A dropped seat keeps position, desaturated chip, 55% name, no timer.
 */
@Composable
fun LudoPlayerPod(
    seatView: LudoSeatView?,
    isSeatAssigned: Boolean,
    active: Boolean,
    ringFraction: Float?,
    ringColorOverride: androidx.compose.ui.graphics.Color?,   // null → player hue
    compact: Boolean,
    modifier: Modifier = Modifier,
) {
    val width = if (compact) LudoDimens.podWidthCompact else LudoDimens.podWidthStandard
    val height = if (compact) LudoDimens.podHeightCompact else LudoDimens.podHeightStandard
    val chip = if (compact) LudoDimens.podChipCompact else LudoDimens.podChipStandard
    val ring = if (compact) LudoDimens.timerRingCompact else LudoDimens.timerRingStandard

    Row(
        modifier = modifier
            .size(width, height)
            .clip(RoundedCornerShape(LudoDimens.podCornerRadius))
            .background(LudoPalette.podSurface()),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.Start,
    ) {
        Box(Modifier.padding(start = (height - ring) / 2)) {
            // Ring surrounds the color chip rather than adding a third ornament (§11.2).
            val darkNow = !com.voiid.app.ui.theme.isLightTheme()
            val resolvedColors = ludoPaletteFor(darkNow)
            PodRing(
                diameter = ring,
                stroke = LudoDimens.timerRingStroke,
                fraction = if (active && ringFraction != null) ringFraction else 1f,
                arcColor = when {
                    !active -> resolvedColors.c(resolvedColors.timerTrack)
                    ringColorOverride != null -> ringColorOverride
                    else -> seatView?.let { resolvedColors.hue(it.seat) }
                        ?: resolvedColors.c(resolvedColors.timerTrack)
                },
                trackColor = resolvedColors.c(resolvedColors.timerTrack),
                chipSize = chip,
                chipColor = seatView?.let { resolvedColors.hue(it.seat) } ?: Color.Transparent,
                outlined = chipOutlined(seatView),
            )
        }
        Text(
            text = usernameLine(seatView),
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
            style = TextStyle(
                fontFamily = androidx.compose.ui.text.font.FontFamily.SansSerif,
                fontWeight = FontWeight.SemiBold,
                fontSize = 13.sp,
            ),
            color = LudoPalette.textPrimary(),
            modifier = Modifier
                .padding(start = 8.dp)
                .weight(1f, fill = false)
                .padding(end = 8.dp),
        )
    }
}

private fun chipOutlined(seatView: LudoSeatView?): Boolean =
    seatView == null || seatView.isWaiting

@Composable
private fun usernameLine(seatView: LudoSeatView?): String = when {
    seatView == null -> ""
    seatView.isWaiting -> "Waiting…"
    else -> (seatView.displayName + if (seatView.isBot) "  BOT" else "").take(22)
}

@Composable
private fun PodRing(
    diameter: Dp,
    stroke: Dp,
    fraction: Float,
    arcColor: androidx.compose.ui.graphics.Color,
    trackColor: androidx.compose.ui.graphics.Color,
    chipSize: Dp,
    chipColor: androidx.compose.ui.graphics.Color,
    outlined: Boolean,
) {
    Box(contentAlignment = Alignment.Center) {
        androidx.compose.foundation.Canvas(modifier = Modifier.size(diameter)) {
            drawCircle(trackColor, style = androidx.compose.ui.graphics.drawscope.Stroke(stroke.toPx()))
            val f = fraction.coerceIn(0f, 1f)
            if (f > 0f) {
                drawArc(
                    color = arcColor,
                    startAngle = -90f,
                    sweepAngle = -360f * f,
                    useCenter = false,
                    style = androidx.compose.ui.graphics.drawscope.Stroke(
                        stroke.toPx(), cap = androidx.compose.ui.graphics.StrokeCap.Round,
                    ),
                )
            }
        }
        Box(
            Modifier
                .size(chipSize)
                .then(
                    if (outlined) Modifier.border(1.dp, LudoPalette.yardPocketBorder(), CircleShape)
                    else Modifier
                )
                .clip(CircleShape)
                .background(chipColor)
        )
    }
}
